workflow/executor/kubelet/kubelet.go

```diff
@@ -0,0 +1,102 @@
+package kubelet
+
+import (
+	"compress/gzip"
+	"os"
+	"syscall"
+	"time"
+
+	"github.com/argoproj/argo/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// killGracePeriod is the time in seconds after sending SIGTERM before
+// forcefully killing the sidecar with SIGKILL (value matches k8s)
+const killGracePeriod = 30
+
+type KubeletExecutor struct {
+	cli *kubeletClient
+}
+
+func NewKubeletExecutor() (*KubeletExecutor, error) {
+	log.Infof("Creating a kubelet executor")
+	cli, err := newKubeletClient()
+	if err != nil {
+		return nil, errors.InternalWrapError(err)
+	}
+	return &KubeletExecutor{
+		cli: cli,
+	}, nil
+}
+
+func (k *KubeletExecutor) GetFileContents(containerID string, sourcePath string) (string, error) {
+	b, err := k.cli.GetFileContents(containerID, sourcePath)
+	if err != nil {
+		return "", err
+	}
+	return b.String(), nil
+}
+
+func (k *KubeletExecutor) CopyFile(containerID string, sourcePath string, destPath string) error {
+	log.Infof("Archiving %s:%s to %s", containerID, sourcePath, destPath)
+	b, err := k.cli.GetFileContents(containerID, sourcePath)
+	if err != nil {
+		return err
+	}
+	f, err := os.OpenFile(destPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0666)
+	if err != nil {
+		return err
+	}
+	w := gzip.NewWriter(f)
+	_, err = w.Write(b.Bytes())
+	if err != nil {
+		return err
+	}
+	err = w.Flush()
+	if err != nil {
+		return err
+	}
+	err = w.Close()
+	if err != nil {
+		return err
+	}
+	return f.Close()
+}
+
+// GetOutput returns the entirety of the container output as a string
+// Used to capturing script results as an output parameter
+func (k *KubeletExecutor) GetOutput(containerID string) (string, error) {
+	return k.cli.GetContainerLogs(containerID)
+}
+
+// Wait for the container to complete
+func (k *KubeletExecutor) Wait(containerID string) error {
+	return k.cli.WaitForTermination(containerID, time.Hour*24)
```

Comment:
I don't think we can pick a timeout that works for everyone. Why wouldn't this just be infinite?

Reply:
Agreed, I made the appropriate changes :+1: